File: 2022/day3/main.py
#!/usr/bin/env python3
from functools import reduce


def find_duplicate(string):
    length = len(string) // 2
    first = set(string[0:length])
    second = set(string[length:])

    return first.intersection(second).pop()


def find_group_badge(string1, string2, string3):
    first = set(string1)
    second = set(string2)
    third = set(string3)

    return first.intersection(second).intersection(third).pop()


def sum_priority_by_group_badge(lines):
    acc = 0
    for i in range(0, len(lines), 3):
        acc += calculate_priority(find_group_badge(lines[i], lines[i + 1], lines[i + 2]))
    return acc


def calculate_priority(char):
    if char == char.lower():
        return ord(char) - 96

    if char == char.upper():
        return ord(char) - 38


with open('input.txt', 'r') as f:
    lines = tuple(f.read().splitlines())

    print(reduce(lambda acc, line: acc + calculate_priority(find_duplicate(line)), lines, 0))
    print(sum_priority_by_group_badge(lines))

File: 2022/day3/main.js
const fs = require('fs')

function findDuplicate(itemListString) {

    const itemList = itemListString.split('')

    const len = itemList.length/2
    const first = new Set(itemListString.substring(0, len).split(''))

    return itemListString.substring(len).split('').filter(el => first.has(el))[0]
}

function findGroupBadge(itemListString1, itemListString2, itemListString3) {

    const first = new Set(itemListString1.split(''))
    const second = new Set(itemListString2.split(''))

    return itemListString3.split('').filter(el => first.has(el) && second.has(el))[0]

}

function calculatePriority(item) {
    if (!item) {
        return 0
    }

    if (item === item.toLowerCase()) {
        return item.charCodeAt(0) - 96
    }

    if (item === item.toUpperCase()) {
        return item.charCodeAt(0) - 38
    }
}

function sumDuplicatePriorities(input) {
    return input.reduce((acc, el) => {
        return acc + calculatePriority(findDuplicate(el))
    }, 0)
}

function sumBadgesPriorities(input) {
    let sum = 0
    for (let i = 0; i <= input.length - 3; i = i + 3) {
        sum = sum + calculatePriority(findGroupBadge(input[i], input[i+1], input[i+2]))
    }
    return sum
}

const input = fs.readFileSync('input.txt', 'utf-8').split('\n')

console.log(sumDuplicatePriorities(input))
console.log(sumBadgesPriorities(input))
